### PR TITLE
PWN-7591 - New Swap. Fix fees in SOL

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7591

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/jupiter/repository/model/JupiterSwapToken.kt [hunk 1]
```diff
@@ -11,4 +12,10 @@ data class JupiterSwapToken(
     val tokenName: String,
     val tokenSymbol: String,
     val tags: List<String>
-)
+) {
+    fun isSol(): Boolean = tokenMint.base58Value == Constants.WRAPPED_SOL_MINT
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/jupiter/repository/model/JupiterSwapToken.kt:16]: I created SolAddress wrapper class, I think we can try to cast it there, and then call isSol fun

File: app/src/main/java/org/p2p/wallet/jupiter/repository/model/JupiterSwapToken.kt [hunk 2]
```diff
@@ -11,4 +12,10 @@ data class JupiterSwapToken(
...
+}
+
+fun List<JupiterSwapToken>.findTokenByMint(mint: Base58String): JupiterSwapToken? {
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/jupiter/repository/model/JupiterSwapToken.kt:19]: why you did not use firstOrNull { }  ?

File: app/src/main/java/org/p2p/wallet/jupiter/repository/tokens/JupiterSwapTokensRemoteRepository.kt [hunk 3]
```diff
@@ -67,4 +72,22 @@ internal class JupiterSwapTokensRemoteRepository(
         val now = System.currentTimeMillis()
         return (now - fetchTokensDate) <= TimeUnit.DAYS.toMillis(1) // check day has passed
     }
+
+    override suspend fun getTokenRate(token: JupiterSwapToken): TokenPrice? =
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/jupiter/repository/tokens/JupiterSwapTokensRemoteRepository.kt:76]: Screenshot 2023-03-12 at 23 29 50
You can unwrap  from dispatcher(io) cause our repository calls already wrapped into dispatchers

**Review by Davran331122 — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt [hunk 4]
```diff
@@ -180,16 +181,22 @@ class JupiterSwapSettingsPresenter(
                 loadingMapper.mapLoadingList()
             }
             is SwapState.LoadingTransaction -> {
+                val solToken = tokens.firstOrNull { it.isSol() }
+                val solTokenRate = solToken?.let { swapTokensRepository.getTokenRate(it) }
```
Comment (NikolaiEmelianov) [inline app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt:185]: loading each time? 
it caching, like was in mapper
```
val tokenRate = swapStateManager.getTokenRate(solToken)
            .filterIsInstance<SwapRateLoaderState.Loaded>()
            .firstOrNull()
            ?.rate
```

**Review by NikolaiEmelianov — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/p2p/wallet/jupiter/repository/model/JupiterSwapToken.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/jupiter/repository/model/JupiterSwapToken.kt:19]: it was used previously in mapper only, this signature makes it more usable then just firstOfNull

File: app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt (hunk 4 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt:185]: getTokenRate works only with SwapTokenModel, not with JupiterSwapToken
i'll check what i can do

File: app/src/main/java/org/p2p/wallet/jupiter/repository/tokens/JupiterSwapTokensRemoteRepository.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/jupiter/repository/tokens/JupiterSwapTokensRemoteRepository.kt:76]: done, thanks!

File: app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt (hunk 4 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt:185]: checked
nothing really can be done here
or cache prices in the repository or rewrite RateManager
i can suggest only to skip this rate request and return to this issue if it becomes one :)

**Review by NikolaiEmelianov — APPROVED** [review]

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt (hunk 4 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt:185]: you can just create  SwapTokenModel.JupiterToken(JupiterSwapToken)

Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/jupiter/ui/settings/presenter/JupiterSwapSettingsPresenter.kt:185]: or load tokens with org.p2p.wallet.swap.jupiter.interactor.SwapTokensInteractor#getAllTokens

**Pull request merged**